x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/ChunkedTrainedModelRestorer.java

```diff
@@ -158,54 +158,58 @@ private void doSearch(
 
-                    boolean continueSearching = modelConsumer.apply(doc);
-                    if (continueSearching == false) {
-                        // signal the search has finished early
-                        successConsumer.accept(Boolean.FALSE);
+                        boolean continueSearching = modelConsumer.apply(doc);
+                        if (continueSearching == false) {
+                            // signal the search has finished early
+                            successConsumer.accept(Boolean.FALSE);
+                            return;
+                        }
+
+                    } catch (IOException e) {
+                        logger.error(() -> "[" + modelId + "] error writing model definition", e);
+                        errorConsumer.accept(e);
                         return;
                     }
-
-                } catch (IOException e) {
-                    logger.error(() -> "[" + modelId + "] error writing model definition", e);
-                    errorConsumer.accept(e);
-                    return;
                 }
-            }
 
-            numDocsWritten += searchResponse.getHits().getHits().length;
+                numDocsWritten += searchResponse.getHits().getHits().length;
 
-            boolean endOfSearch = searchResponse.getHits().getHits().length < searchSize
-                || searchResponse.getHits().getTotalHits().value == numDocsWritten;
+                boolean endOfSearch = searchResponse.getHits().getHits().length < searchSize
+                    || searchResponse.getHits().getTotalHits().value == numDocsWritten;
 
-            if (endOfSearch) {
-                successConsumer.accept(Boolean.TRUE);
-            } else {
-                // search again with after
-                SearchHit lastHit = searchResponse.getHits().getAt(searchResponse.getHits().getHits().length - 1);
-                SearchRequestBuilder searchRequestBuilder = buildSearchBuilder(client, modelId, index, searchSize);
-                searchRequestBuilder.searchAfter(new Object[] { lastHit.getIndex(), lastNum });
-                executorService.execute(() -> doSearch(searchRequestBuilder.request(), modelConsumer, successConsumer, errorConsumer));
+                if (endOfSearch) {
+                    successConsumer.accept(Boolean.TRUE);
+                } else {
+                    // search again with after
+                    SearchHit lastHit = searchResponse.getHits().getAt(searchResponse.getHits().getHits().length - 1);
+                    SearchRequestBuilder searchRequestBuilder = buildSearchBuilder(client, modelId, index, searchSize);
+                    searchRequestBuilder.searchAfter(new Object[] { lastHit.getIndex(), lastNum });
+                    executorService.execute(() -> doSearch(searchRequestBuilder.request(), modelConsumer, successConsumer, errorConsumer));
+                }
+            } finally {
+                searchResponse.decRef();
```

Comment:
Since this recurses in `executorService.execute(() -> doSearch(searchRequestBuilder.request(), modelConsumer, successConsumer, errorConsumer));` I am not sure we want to delay the decRef until all the recursed calls are finished, no?